core/src/main/scala/org/bitcoins/core/crypto/MnemonicCode.scala

```diff
@@ -0,0 +1,247 @@
+package org.bitcoins.core.crypto
+
+import java.security.SecureRandom
+
+import org.bitcoins.core.protocol.NetworkElement
+import org.bitcoins.core.util.{CryptoUtil, Factory}
+import scodec.bits.{BitVector, ByteVector}
+
+import scala.annotation.tailrec
+import scala.io.Source
+
+sealed abstract class MnemonicCode extends NetworkElement {
+
+  private val VALID_LENGTHS = Vector(12, 15, 18, 21, 24)
+  require(
+    VALID_LENGTHS.contains(words.length),
+    // $COVERAGE-OFF$
+    s"Number of words must be one of the following: ${VALID_LENGTHS.mkString(", ")} "
+    // $COVERAGE-ON$
+  )
+
+  require({
+    val entropy = toEntropyWithChecksum
+    isEntropyWithChecksumValid(entropy)
+  }, "Entropy checksum is not valid!")
+
+  /**
+    * Checks that the provided entropy has a valid checksum
+    * attached at the end
+    */
+  private[crypto] def isEntropyWithChecksumValid(
+      entropyWithChecksum: BitVector): Boolean = {
+
+    val codeInfo = MnemonicCode.getMnemonicCodeInfo(words)
+    val entropyNoChecksum = entropyWithChecksum.take(codeInfo.entropyBits)
+
+    val hashedEntropy = CryptoUtil.sha256(entropyNoChecksum).bytes.toBitVector
+    val checksum = hashedEntropy.take(codeInfo.checksumLength)
+
+    entropyNoChecksum ++ checksum == entropyWithChecksum
+
+  }
+
+  def words: Vector[String]
+  override def bytes: ByteVector = ByteVector(words.mkString(" ").getBytes)
+
+  /**
+    * Returns the entropy initially provided to construct
+    * this mnemonic code
+    */
+  def toEntropy: BitVector = {
+    val entropyWithChecksumBits = toEntropyWithChecksum
+    val lengthNoEntropy = MnemonicCode
+      .getMnemonicCodeInfo(words)
+      .entropyBits
+
+    entropyWithChecksumBits.take(lengthNoEntropy)
+  }
+
+  /**
+    * Returns the entropy _with checksum_ originally provided
+    * to construct this mnemonic code
+    */
+  def toEntropyWithChecksum: BitVector = {
+
+    def bitsForWord(word: String): BitVector = {
+
+      // tail-recursive binary search for word in word list
+      // stolen from https://codereview.stackexchange.com/questions/161887/re-inventing-the-wheel-binary-search-in-scala
+      import MnemonicCode.{ENGLISH_WORDS => wordlist}
+      @tailrec
+      def search(
```

Comment:
Any chance we have some place for utility functions like this in core?

Reply:
Haven't found one. Can add it to `BitcoinSUtil`? What would the signature look like, in that case? 

What I tried: 

```scala
def binarySearch[
      T <: Ordered[T],
  ](sortedCollection: Vector[T], elem: T): Option[Int] 
```

the method itself compiles, but not passing it a `Vecto[String]` and a `String`